### PR TITLE
Add validations for swap fields

### PR DESCRIPTION
**Pull request by antonijzelinskij**
#8692ty65x

### CODE REVIEW

**Comment by stepanLav** [comment]
Task linked: [CU-8692ty65x Android. Validations on Continue](https://app.clickup.com/t/8692ty65x)

**Comment by stepanLav** [comment]
Task linked: [CU-8692ty65x Android. Validations on Continue](https://app.clickup.com/t/8692ty65x)

**Comment by stepanLav** [comment]
Task linked: [CU-8692ty65x Android. Validations on Continue](https://app.clickup.com/t/8692ty65x)

File: common/src/main/java/io/novafoundation/nova/common/validation/FieldValidator.kt [hunk 1]
```diff
@@ -19,6 +20,23 @@ abstract class MapFieldValidator : FieldValidator {
...
 }
 
+class CompoundFieldValidator(
+    private val validators: List<FieldValidator>
+) : FieldValidator {
+
+    constructor(vararg validators: FieldValidator) : this(validators.toList())
+
+    override fun observe(inputStream: Flow<String>): Flow<FieldValidationResult> {
+        // The negative side of this approach is that we run all validations despite one of them may be failed
```
Comment (valentunn) [inline common/src/main/java/io/novafoundation/nova/common/validation/FieldValidator.kt:30]: I think its fine given that validation produce not a single value but a stream of values

File: feature-swap-impl/src/main/java/io/novafoundation/nova/feature_swap_impl/presentation/main/SwapMainSettingsViewModel.kt [hunk 2]
```diff
@@ -163,30 +166,39 @@ class SwapMainSettingsViewModel(
...
 
-    val showDetails: Flow<Boolean> = quotingState.map { it !is QuotingState.NotAvailable }
+    val showDetails: Flow<Boolean> = quotingState.map { it !is QuotingState.NotAvailable && it !is QuotingState.Default }
         .shareInBackground()
 
     private val _validationProgress = MutableStateFlow(false)
 
-    val buttonState: Flow<DescriptiveButtonState> = flowOf { DescriptiveButtonState.Enabled(resourceManager.getString(R.string.common_continue)) }
+    val buttonState: Flow<DescriptiveButtonState> = combine(
+        accumulate(amountInInput.fieldError, amountOutInput.fieldError),
```
Comment (valentunn) [inline feature-swap-impl/src/main/java/io/novafoundation/nova/feature_swap_impl/presentation/main/SwapMainSettingsViewModel.kt:195]: Just a side note  on `accumulate`. Here is our current implementation:
image
I've recently realized we can utilize `runningFold` to achieve much cleaner version. For example, here is the code I've written for maps:
image
The only side effect of such refactoring is that ordering of original list might not be preserved

File: feature-swap-impl/src/main/java/io/novafoundation/nova/feature_swap_impl/presentation/main/SwapMainSettingsViewModel.kt [hunk 3]
```diff
@@ -163,30 +166,39 @@ class SwapMainSettingsViewModel(
...
+        assetOutFlow,
+        amountInInput.inputState,
+        amountOutInput.inputState
+    ) { fieldErrorStates, assetOut, amountIn, amountOut ->
+        formatButtonStates(fieldErrorStates, assetOut, amountIn, amountOut)
```
Comment (valentunn) [inline feature-swap-impl/src/main/java/io/novafoundation/nova/feature_swap_impl/presentation/main/SwapMainSettingsViewModel.kt:200]: Can it be passed as a reference - `::formatButtonStates`?

**Review by valentunn — APPROVED** [review]

**Pull request merged**